### ISSUE TITLE
VS Code breakpoint with tsx and Nodejs 20.12.0 does not work

### ISSUE THREAD

**Issue by piquark6046** [privatenumber/tsx#506]
### Acknowledgements

- [X] I searched existing issues to avoid duplicates
- [X] I understand this is a place to report a confirmed bug (not seek debugging help)
- [X] I understand this is a collaborative open source project, and relies on community contributions
- [X] I have read and understood the [Contribution guide](https://github.com/privatenumber/tsx/blob/develop/CONTRIBUTING.md)

### Minimal reproduction URL

https://github.com/List-KR/multithread-array/pull/1

### Version

4.7.1

### Node.js version

20.12.0

### Package manager

npm

### Operating system

Linux

### Problem & expected behavior (under 200 words)

- What happened: All VS Code breakpoints are ignored
- What I expected: VS Code breakpoint works as using Node.js 20.11.1
- Screenshot:
  <details><summary>Screenshot</summary>
  <p>
  
  ![image](https://github.com/privatenumber/tsx/assets/98787049/8ee25c54-f699-4b82-a263-8782b1d63847)
  
  
  </p>
  </details> 

- devcontainer.json
  ```json
  {
	  "name": "Typescript with Rust",
	  "image": "mcr.microsoft.com/devcontainers/base:ubuntu",
	  "features": {
		  "ghcr.io/devcontainers/features/node:latest": {
			  "nodeGypDependencies": true,
			  "version": "lts"
		  },
		  "ghcr.io/devcontainers/features/rust:latest": {
			  "version": "latest",
			  "profile": "default"
		  },
		  "ghcr.io/devcontainers-contrib/features/bash-command:latest": {
			  "command": "echo hi!"
		  },
		  "ghcr.io/devcontainers-contrib/features/npm-package:latest": {
			  "package": "typescript",
			  "version": "latest"
		  },
		  "ghcr.io/devcontainers-contrib/features/typescript:latest": {
			  "version": "latest"
		  },
		  "devwasm.azurecr.io/dev-wasm/dev-wasm-feature/rust-wasi:latest": {}
	  },
	  "remoteUser": "root",
	  "mounts": [
		  "source=${localEnv:HOME}/repos/Filters,target=/workspaces/resources/Filters,type=bind,consistency=cached",
		  "source=${localEnv:HOME}/repos/Userscript,target=/workspaces/resources/Userscript,type=bind,consistency=cached"
	  ]
  }
  ```

### Contributions

- [ ] I will open a pull request for this issue
- [ ] I will support the work financially (any amount helps)

**Comment by piquark6046** [comment]
Is The reproduction information not enough?

**Comment by privatenumber** [comment]
Took a glance and the first few pages I scrolled had nothing to do with tsx or VSCode.

**Comment by tenpaMk2** [comment]
Hi @privatenumber . Thank you for developing.

I can reproduce in the minimal project.

- [tsx-vscode-debug-not-working](https://github.com/tenpaMk2/tsx-vscode-debug-not-working)

VSCode cannot stop at the breakpoints via Node 20.12.0.

However, it can do via Node 20.11.1.
So, I think it's a problem of Node.

- Mac OS: Sonoma 14.4.1
- VSCode: February 2024 (version 1.87)
- Node: 20.12.0
- tsx: 4.7.1

Both Node 21 and 20 are not working at tsx when debugging...
I think it's too difficult situation for tsx users...

**Comment by piquark6046** [comment]
@privatenumber VS Code breakpoint works with Node.js 18.20.0.
However, it does work with Node.js 20.12.0.
Please take a look. I think that it is a major bug of your tsx package.

**Comment by piquark6046** [comment]
<details><summary>Reproduction video</summary>
<p>

[Screencast from 2024-04-02 01-29-11.webm](https://github.com/privatenumber/tsx/assets/98787049/89febf89-f0f0-4301-a87a-7e5119644a66)


</p>
</details> 

<details><summary>devcontainer.json (Docker Engine rootless)</summary>
<p>

```json
{
    "name": "Typescript",
    "image": "mcr.microsoft.com/devcontainers/base:ubuntu",
    "features": {
  	  "ghcr.io/devcontainers/features/node:latest": {
  		  "nodeGypDependencies": true,
  		  "version": "lts"
  	  },
  	  "ghcr.io/devcontainers-contrib/features/bash-command:latest": {
  		  "command": "echo hi!"
  	  },
  	  "ghcr.io/devcontainers-contrib/features/npm-package:latest": {
  		  "package": "typescript",
  		  "version": "latest"
  	  },
  	  "ghcr.io/devcontainers-contrib/features/typescript:latest": {
  		  "version": "latest"
  	  }
    },
    "remoteUser": "root"
}
```

</p>
</details>

**Comment by piquark6046** [comment]
One day,

**Comment by piquark6046** [comment]
Sure. What the bug does not affect you can be happened.
However, one day, you may be affected if Node.js 18.x become end-of-life.
Also, please remind dependency impact of your tsx package to npm ecosystem.

**Comment by piquark6046** [comment]
Paid 15 USD.

**Comment by aslampr07** [comment]
Exact same problem here, downgraded `21.7.1` to `20.11.1` and it is working again.

**Comment by privatenumber** [comment]
@piquark6046 

This is a project I work on for fun, primarily to improve my own workflow.

I'm not incentivized to fix this now (maybe in the future). But if you're affected by it, it makes sense for you to fix it and demonstrate how much you care about the npm ecosystem.

I already donate my time to open source and maintain tsx to help others, and I'm not trying to do more work for free.

I'd be happy to review and help you with your PR for free though.

**Comment by cenfun** [comment]
It seems that the root cause is because inline sourcemap no longer works in nodejs 20.12.0. VSCode debug might not work if there is no inline sourcemap.
-  `(typeof Error.prepareStackTrace !== "function") === false` in nodejs 20.12.0, (see change https://github.com/nodejs/node/pull/50827/files) but it is `true` in nodejs 20.11.0
- `tsx` used above to install inline sourcemap, see https://github.com/privatenumber/tsx/blob/develop/src/source-map.ts#L21

Hope this helps you.

**Comment by privatenumber** [comment]
Ah very helpful @cenfun, great work!

Would you be willing to open a PR?

**Comment by piquark6046** [comment]
@privatenumber 
Please note that the reason I paid is to request resolving this issue.

**Comment by privatenumber** [comment]
@piquark6046

If you'd like Priority Support, please see the tiers here: https://github.com/sponsors/privatenumber

**Comment by bqp-articulate** [comment]
I can confirm this behavior, and I'll add that I can also reproduce with the `tsx` smoketests in 20.12.  If @cenfun isn't submitting a PR for this, I'd be happy to - I've invested time already wrapping my head around this bit of code.

**Comment by cenfun** [comment]
@bqp-articulate there is PR here https://github.com/privatenumber/tsx/pull/517

**Comment by cenfun** [comment]
@privatenumber  I create a new PR here https://github.com/privatenumber/tsx/pull/518
Could you please review it?

**Comment by piquark6046** [comment]
@privatenumber Can you tell me when the fixes is released via npm?

**Comment by privatenumber** [comment]
Released in v4.7.2!

Thanks for your contribution @cenfun 🙌!
